[BE][Team-17][Robin, Cooper] 반찬가게 1차 PR 요청드립니다.

pbg0205 opened the pull request:
안녕하세요. team17의 백엔드 로빈, 쿠퍼입니다.
반찬가게 API 구현 1차 PR 드립니다.😄
아래에 구현 목록 사항과 목업 API 명세서 URL 첨부하여 보내드립니다.
부족한 코드, 시간내어 리뷰해주셔서 감사합니다. 🙇‍♂️ 🙇‍♀️

### 구현 목록 사항
- 베스트 카테고리 기준 전체 상품 목록 조회

- 베스트 카테고리 기준 세부 상품 목록 조회

- 밥찬들 카테고리 기준 세부 상품 목록 조회


### [목업 API 명세서 URL](https://documenter.getpostman.com/view/8052286/TzJuBe2M)

pbg0205 committed:
Chore : 백엔드 프로젝트 파일을 생성하라.

malaheaven committed:
Feat: 스프링 데이터 jdbc으로 셋팅하라

- 스프링 부트와 mysql을 연동해라
- 반찬 item, image, best_category, dish_category 테이블을 생성해라

pbg0205 committed:
Merge pull request #10 from swing-park/feature-setup-jdbc

Feat: 스프링 데이터 jdbc으로 셋팅하라

malaheaven committed:
Feat: domain을 생성해라

- Item domain을 생성해라
- Image domain을 생성해라
- DishCategory domain을 생성해라
- BestCategory domain을 생성해라

pbg0205 committed:
Merge pull request #11 from swing-park/feature-create-domain

Feat: domain을 생성해라

malaheaven committed:
Feat: repository를 생성해라

- ItemRepository를 생성해라
- ImageRepository를 생성해라
- DishCategoryRepository를 생성해라
- BestCategoryRepository를 생성해라

pbg0205 committed:
Merge pull request #12 from swing-park/feature-create-repository

Feat: repository를 생성해라

malaheaven committed:
Fix : DB 스키마 변경하라

- 변경된 image, item 컬럼에 따라 domain 변경
- best_category 및 dish_category 에 대한 insert구문 추가

pbg0205 committed:
Merge pull request #17 from swing-park/fix-db-schema

Fix : DB 스키마 변경하라

malaheaven committed:
Feat: controller를 만들어라

pbg0205 committed:
Merge pull request #18 from swing-park/feature-create-controller

Feat: controller를 만들어라

pbg0205 committed:
style : 변수명 오탈자를 수정하라.

- discription -> description으로 변수명을 변경하라.

pbg0205 committed:
feat : finAll 메서드를 추가하라.

- 리턴타입을 List로 한 findAll 메서드를 생성하라.

pbg0205 committed:
feat : domain과 dto에 getter를 추가해라.

pbg0205 committed:
feat : 베스트목록들을 반환하는 메서드를 작성하라.

pbg0205 committed:
feat : 베스트 카테고리 목록을 조회하라.

- 베스트 카테고리 전체 목록을 조회하라.

- 베스트 카테고리 세부 목록을 조회하라.

pbg0205 committed:
feat : service layer를 추가하라.

- ItemService를 추가하라.

- ImageService를 추가하라.

- BestCategoryService를 추가하라.

malaheaven committed:
Merge pull request #21 from swing-park/feature-create-get-method

Feature create get method

pbg0205 committed:
style : ItemDto item 변수명을 변경하라.

- itemId -> item_id로 변경하라.

malaheaven committed:
Merge pull request #22 from swing-park/feature-create-get-method

style : ItemDto item 변수명을 변경하라.

malaheaven committed:
Feat: DishCategory 조회 기능을 구현해라

- DishCategoryController 구현
- DishCategoryService 구현
- DishCategoryDto 구현

pbg0205 committed:
Merge pull request #24 from swing-park/feature-create-dishcategory

Feat: DishCategory 조회 기능을 구현해라

wheejuni commented on backend/sidedish/src/main/java/codsquad/team17/sidedish/controller/BestCategoryController.java:
```diff
@@ -0,0 +1,29 @@
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/best")
```
클라이언트 개발자들과 협의한 거겠지만 그리 좋은 path로 보이진 않습니다.
API인지도 불분명하고, 무엇보다 무엇을 위한 best인지 잘 나타나지 않는 것 같아요.

wheejuni commented on backend/sidedish/src/main/java/codsquad/team17/sidedish/controller/DishCategoryController.java:
```diff
@@ -0,0 +1,39 @@
+public class DishCategoryController {
+
+    private final DishCategoryService dishCategoryService;
+    private final ItemService itemService;
+    private final ImageService imageService;
```
우선 이 두 서비스는 아직 사용되지 않는 것 같네요.
그리고 하나의 컨트롤러에 세 개의 서비스나 필요하다면, 컨트롤러가 분리돼야 하는 건 아닌지 고민해 보셔야 할 수도 있겟습니다. 
이미지와 아이템까지 여기서 처리해야 하나요?

pbg0205 replied:
리팩토링 과정에서 삭제하지 못한 의존성 주입입니다ㅠ 변경하도록 하겠습니다!!

wheejuni commented on backend/sidedish/src/main/java/codsquad/team17/sidedish/domain/Item.java:
```diff
@@ -0,0 +1,75 @@
+    private String description;
+
+    private String badge;
+    private BigDecimal nPrice;
+
+    private BigDecimal sPrice;
```
`BigDecimal` 사용 좋습니다. 👍 
그러나 `nPrice`, `sPrice` 라는 변수명은 아쉽네요. `n` 과 `s` 에 담긴 의미가 잘 유추되지 않습니다.

pbg0205 replied:
MockAPI 변수 그대로 사용하는 과정에서 변수명 의미가 불명확하게 선언했네요ㅠ
normalPrice와 salePrice로 변수명 변경하도로 하겠습니다!
피드백 주셔서 감사합니다.😃

wheejuni commented on backend/sidedish/src/main/java/codsquad/team17/sidedish/dto/ItemDto.java:
```diff
@@ -0,0 +1,65 @@
+    private String title;
+    private String description;
+
+    private BigDecimal n_price;
+    private BigDecimal s_price;
```
어떤 경우에도 Java의 코딩 컨벤션에 맞지 않는 변수명을 사용하는 것은 자제해 주시면 좋겠습니다.
만약 JSON response를 위한 커스텀 필드명을 사용하고 싶다면, 아래 링크를 참조해 적절한 애노테이션을 이용하면 됩니다. 

https://www.baeldung.com/jackson-annotations

만약 카멜 케이스 변수명에 대해 일괄적으로 스네이크 케이스를 사용하는 것이 목적이라면, 아래 링크를 보고 학습해 보시는 것도 좋겠습니다. (아래 링크에 담긴 첫번째 솔루션은 권장하지 않습니다)

https://stackoverflow.com/questions/10519265/jackson-overcoming-underscores-in-favor-of-camel-case

pbg0205 replied:
위에 참고해주신 내용을 바탕으로 한번 수정해보도록 하겠습니다!

wheejuni commented on backend/sidedish/src/main/java/codsquad/team17/sidedish/service/BestCategoryService.java:
```diff
@@ -0,0 +1,45 @@
+public class BestCategoryService {
+    private final BestCategoryRepository bestCategoryRepository;
+
+    private final ItemService itemService;
+    private final ImageService imageService;
```
서비스에서 타 서비스에 의존하는 방식은 가급적이면 피해 주시고,
지금 객체 그래프 탐색이 되지않다보니 이 두 엔티티를 조회하는 것이 불가피한 측면이 있어보이는데요,
`ItemRepository` 와 `ImageRepository` 를 바로 주입받는 방향이 더 좋아 보입니다.

wheejuni reviewed (changes requested):
수고 많으셨습니다. 👍 
이번 프로젝트는 스펙 규모가 꽤 되는 거 같네요. 많은 고민 끝에 작성하신 코드 잘 보았습니다.

개선하면 좋을 부분 추려서 피드백 남겨보았는데 확인 부탁드릴게요.

pbg0205 committed:
refactor : 불필요한 의존성을 제거하라.

- 사용하지 않는 ItemService 제거

- 사용하지 않는 ImageService 제거

pbg0205 committed:
feat : BestCategory의 불필요한 의존성을 제거하라.

- 사용하지 않는 ImageService를 제거하라.

pbg0205 committed:
refactor : imageRepository 의존성을주입하자.

- 현재 단계에서 imageService의 큰 사용이 없어 조회기능을
imageRepository로 주입해서 조회하도록 변경하라.

pbg0205 committed:
refactor : dto의 nPrice, sPrice 변수명을 변경하라.

- Java Naming Convention에 맞게 camel case로 변경하여라.

- 정확한 의미를 부여할 수 있도록 변수명을 구체적으로 작성하라.

pbg0205 committed:
refactor : itemDto 변수명에 final을 선언하라.

pbg0205 committed:
feat : 상세정보를 조회하는 기능을 구현하라.

malaheaven committed:
Merge pull request #29 from swing-park/feature-create-detail

Feature create detail

pbg0205 committed:
refactor : API관련 내용을 수정하라.

- itemDetail : description 2500원 내용 삭제

- badge 빈 문자열 제거
- badge 베스트 요소 제거

- 배포 URL 완성
- URL 추가 : recommandedItemDto, ItemDto

- delivery-type 순서 변경

malaheaven committed:
Merge pull request #36 from swing-park/refactor-response-json-api

refactor : API관련 내용을 수정하라.

pbg0205 committed:
feat : CORS 기능을 구현하라.

malaheaven committed:
Merge pull request #37 from swing-park/feature-setup-cors

feat : CORS 기능을 구현하라.

pbg0205 committed:
style : URL 추가 / 제거

malaheaven committed:
Merge pull request #39 from swing-park/style-url-path

style : URL 추가 / 제거

pbg0205 committed:
fix : 오탈자 수정 및 URL 추가 수정

- 중복된 URL을 제거하라.

- recommanded -> recommended로 변경하라.

malaheaven committed:
Merge pull request #41 from swing-park/style-url-path

fix : 오탈자 수정 및 URL 추가 수정

pbg0205 committed:
style : 중복된 URL을 제거하라.

pbg0205 committed:
Merge pull request #43 from swing-park/style-url-path

style : 중복된 URL을 제거하라.

malaheaven committed:
Fix: dish url

malaheaven committed:
Merge pull request #45 from swing-park/fix-dish-url

Fix: dish url

pbg0205 committed:
feat : 주문 기능을 위한 OrderDto 생성

pbg0205 committed:
feat : OrderController 생성

pbg0205 committed:
feat : Item entity에 재고 반영 기능 추가

pbg0205 committed:
feat : 재고 반영을 위한 OrderService를 생성하라.

malaheaven committed:
Merge pull request #50 from swing-park/feature-order-test

[BE] feature order

pbg0205 committed:
feat : ItemDetailDto deliveryInfo 변수명을 추가하라.

pbg0205 committed:
style : dummy data를 수정하라.

- 이벤트 특가, 론칭 특가 순서 변경

pbg0205 committed:
feat : 추천 목록을 10개를 임의로 반환하라.

malaheaven committed:
Merge pull request #51 from swing-park/refactor-FE-feedback

[BE] Refactor FE feedback

pbg0205 committed:
refactor : RequestBody 형태로 데이터를 전달받아라.

malaheaven committed:
Merge pull request #55 from swing-park/feature-order

refactor : RequestBody 형태로 데이터를 전달받아라.

pbg0205 commented:
리뷰해주신 부분에 대해서 수정 완료하였습니다!
급하게 프로젝트 진행하는 과정에서 리뷰 사항 외에 추가 기능 구현 내용이 반영되었습니다ㅠ
이 점 유의하셔서 확인 부탁드립니다.
다시 한번 시간내서 리뷰해주신 점 감사드립니다. 😄

malaheaven committed:
Feat: 커스텀 이셉션 핸들러를 이용하여 예외를 처리

- best 카테고리
- dish 카테고리
- 이미지
- 아이템
- 재고부족

pbg0205 committed:
Merge pull request #61 from swing-park/feature-exception-handler

Feat: 커스텀 이셉션 핸들러를 이용하여 예외를 처리

pbg0205 committed:
feat : CORS 설정에 PUT METHOD를 추가하라.

malaheaven committed:
Merge pull request #62 from swing-park/feat-cors-put

feat : CORS 설정에 PUT METHOD를 추가하라.

pbg0205 committed:
chore : jackson databind dependency를 추가하라.

malaheaven committed:
Merge pull request #66 from swing-park/chore-jackson-core

chore : jackson databind dependency를 추가하라.
